hibernate-core/src/main/java/org/hibernate/loader/ast/internal/CompoundNaturalIdLoader.java

```diff
@@ -0,0 +1,98 @@
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.loader.ast.internal;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.loader.NaturalIdPostLoadListener;
+import org.hibernate.loader.NaturalIdPreLoadListener;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+import org.hibernate.metamodel.mapping.SingularAttributeMapping;
+import org.hibernate.metamodel.mapping.internal.CompoundNaturalIdMapping;
+import org.hibernate.metamodel.mapping.internal.MappingModelCreationProcess;
+import org.hibernate.sql.ast.Clause;
+import org.hibernate.sql.ast.tree.expression.JdbcParameter;
+import org.hibernate.sql.exec.internal.JdbcParameterBindingImpl;
+import org.hibernate.sql.exec.spi.JdbcParameterBindings;
+
+/**
+ * NaturalIdLoader implementation for compound natural-ids
+ */
+public class CompoundNaturalIdLoader<T> extends AbstractNaturalIdLoader<T> {
+
+	public CompoundNaturalIdLoader(
+			CompoundNaturalIdMapping naturalIdMapping,
+			NaturalIdPreLoadListener preLoadListener,
+			NaturalIdPostLoadListener postLoadListener,
+			EntityMappingType entityDescriptor,
+			MappingModelCreationProcess creationProcess) {
+		super( naturalIdMapping, preLoadListener, postLoadListener, entityDescriptor, creationProcess );
+	}
+
+	@Override
+	protected Object resolveNaturalIdBindValue(Object naturalIdValue, SharedSessionContractImplementor session) {
+		// the "real" form as an array, although we also accept here a Map and reduce it to
+		// the appropriately ordered array
+		if ( naturalIdValue instanceof Object[] ) {
+			return naturalIdValue;
+		}
+
+		final List<SingularAttributeMapping> attributes = naturalIdMapping().getNaturalIdAttributes();
+		final Object[] naturalId = new Object[ attributes.size() ];
+
+		if ( naturalIdValue instanceof Map ) {
+			final Map<String,?> valueMap = (Map<String,?>) naturalIdValue;
+			for ( int i = 0; i < attributes.size(); i++ ) {
+				final SingularAttributeMapping attributeMapping = attributes.get( i );
+				naturalId[ i ] = valueMap.get( attributeMapping.getAttributeName() );
+			}
+			return naturalId;
+		}
+
+		throw new IllegalArgumentException( "Unexpected natural-id reference [" + naturalIdValue + "; expecting array or Map" );
+	}
+
+	@Override
+	protected void applyNaturalIdAsJdbcParameters(
+			Object naturalIdToLoad,
+			List<JdbcParameter> jdbcParameters,
+			JdbcParameterBindings jdbcParamBindings,
+			SharedSessionContractImplementor session) {
+		assert naturalIdToLoad instanceof Object[];
+		final Object[] naturalIdValueArray = (Object[]) naturalIdToLoad;
+
+		final Iterator<JdbcParameter> jdbcParamItr = jdbcParameters.iterator();
+
+		for ( int i = 0; i < naturalIdMapping().getNaturalIdAttributes().size(); i++ ) {
+			final SingularAttributeMapping attrMapping = naturalIdMapping().getNaturalIdAttributes().get( i );
```

Comment:
Is that true that we don't need to avoid repeated invoking "naturalIdMapping().getNaturalIdAttributes()" simply because it is too light-weight? Why not ensure we only invoke it only once as:
```
for ( SingularAttributeMapping attrMapping : naturalIdMapping().getNaturalIdAttributes()) {

```

Reply:
At least it hurts code readability, IMHO.

Reply:
I believe we discussed this one before... the general case... enhanced-for-loops are not efficient.  `#forEach` *might be* efficient (depends on a few things).

Reply:
Sorry for the confusion. My point here is the duplication avoidance. Obviously, Java compiler won't be smart enough to change:
```
for ( int i = 0; i < naturalIdMapping().getNaturalIdAttributes().size(); i++ ) {
			final SingularAttributeMapping attrMapping = naturalIdMapping().getNaturalIdAttributes().get( i );

```
into more efficient
```
var naturalIdAttrs = naturalIdMapping().getNaturalIdAttributes();
for ( int i = 0; i < naturalIdAttrs.size(); i++ ) {
			final SingularAttributeMapping attrMapping = naturalIdAttrs.get( i );

```
for the following reasons:

1. the loop body might change the natural id attributes collection;
2. naturalIdMapping().getNaturalIdAttributes() might have side effect, for instance, the common lazy initialization pattern.
The avoidance of duplication leads to better code readability also, I believe.